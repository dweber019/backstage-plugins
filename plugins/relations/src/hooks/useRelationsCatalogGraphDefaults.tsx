import { configApiRef, useApi } from '@backstage/core-plugin-api';
import {
  ALL_RELATION_PAIRS,
  EntityRelationsGraphProps,
} from '@backstage/plugin-catalog-graph';

export function useRelationsCatalogGraphDefaults(
  props: Partial<EntityRelationsGraphProps>,
) {
  const configApi = useApi(configApiRef);

  let modifiedProps = props;
  if (modifiedProps.unidirectional === undefined) {
    modifiedProps = { ...modifiedProps, unidirectional: false };
  }

  if (modifiedProps.relationPairs === undefined) {
    const relations = configApi.getConfigArray('relationsProcessor.relations');
    let finalRelationPairs = ALL_RELATION_PAIRS;
    if (Array.isArray(relations)) {
      finalRelationPairs = [
        ...ALL_RELATION_PAIRS,
        ...relations
          .map(relationConfig =>
            relationConfig
              .getConfigArray('pairs')
              .map(
                pair =>
                  [pair.getString('incoming'), pair.getString('outgoing')] as [
                    string,
                    string,
                  ],
              )
              .filter(
                pair =>
                  !ALL_RELATION_PAIRS.some(
                    standardPair =>
                      (standardPair[0] === pair[0] &&
                        standardPair[1] === pair[1]) ||
                      (standardPair[0] === pair[1] &&
                        standardPair[1] === pair[0]),
                  ),
              ),
          )
          .flat(),
      ];
    }
    modifiedProps = { ...modifiedProps, relationPairs: finalRelationPairs };
  }

  if (modifiedProps.showArrowHeads === undefined) {
    modifiedProps = { ...modifiedProps, showArrowHeads: true };
  }

  return modifiedProps;
}
