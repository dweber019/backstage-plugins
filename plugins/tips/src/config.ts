import { createApiRef, IdentityApi } from '@backstage/core-plugin-api';
import React from 'react';
import { Entity } from '@backstage/catalog-model';

export const tipsConfigRef = createApiRef<TipsConfig>({
  id: 'plugin.tips.config',
});

/** @public */
export interface TipsConfig {
  tips: Tip[];
}

/** @public */
export interface Tip {
  content: string | React.ReactElement;
  title: string;
  activate: (options: { entity?: Entity, identity?: IdentityApi }) => Promise<boolean>;
}
