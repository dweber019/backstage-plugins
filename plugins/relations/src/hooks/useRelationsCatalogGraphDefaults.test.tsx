import { configApiRef } from '@backstage/core-plugin-api';
import { renderHook } from '@testing-library/react';
import { ConfigReader } from '@backstage/config';
import { useRelationsCatalogGraphDefaults } from './useRelationsCatalogGraphDefaults';
import { TestApiProvider } from '@backstage/test-utils';

describe('useRelationsCatalogGraphDefaults', () => {
  const config = new ConfigReader({
    relationsProcessor: {
      relations: [
        {
          sourceKind: 'Component',
          attribute: 'test',
          pairs: [
            {
              incoming: 'testOf',
              outgoing: 'testBy',
            },
          ],
        },
        {
          sourceKind: 'Component',
          attribute: 'owner',
          pairs: [
            {
              incoming: 'ownerOf',
              outgoing: 'ownedBy',
            },
          ],
        },
      ],
    },
  });

  const AllTheProviders = ({ children }: any) => {
    return (
      <TestApiProvider apis={[[configApiRef, config]]}>
        {children}
      </TestApiProvider>
    );
  };

  test('should return arrow=true as default', () => {
    const { result } = renderHook(() => useRelationsCatalogGraphDefaults({}), {
      wrapper: AllTheProviders,
    });

    expect(result.current.showArrowHeads).toBeTruthy();
  });

  test('should return arrow overwrite', () => {
    const { result } = renderHook(
      () => useRelationsCatalogGraphDefaults({ showArrowHeads: false }),
      { wrapper: AllTheProviders },
    );

    expect(result.current.showArrowHeads).toBeFalsy();
  });

  test('should return unidirectional=false as default', () => {
    const { result } = renderHook(() => useRelationsCatalogGraphDefaults({}), {
      wrapper: AllTheProviders,
    });

    expect(result.current.unidirectional).toBeFalsy();
  });

  test('should return unidirectional overwrite', () => {
    const { result } = renderHook(
      () => useRelationsCatalogGraphDefaults({ unidirectional: true }),
      { wrapper: AllTheProviders },
    );

    expect(result.current.unidirectional).toBeTruthy();
  });

  test('should return relationPairs defaults', () => {
    const { result } = renderHook(() => useRelationsCatalogGraphDefaults({}), {
      wrapper: AllTheProviders,
    });

    expect(result.current.relationPairs).toMatchSnapshot();
  });

  test('should return relationPairs overwrite', () => {
    const { result } = renderHook(
      () =>
        useRelationsCatalogGraphDefaults({
          relationPairs: [['overwriteOf', 'overwriteBy']],
        }),
      { wrapper: AllTheProviders },
    );

    expect(result.current.relationPairs).toMatchSnapshot();
  });
});
