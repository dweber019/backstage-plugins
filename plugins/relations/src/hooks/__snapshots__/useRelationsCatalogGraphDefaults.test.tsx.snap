// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRelationsCatalogGraphDefaults should return relationPairs defaults 1`] = `
[
  [
    "ownerOf",
    "ownedBy",
  ],
  [
    "consumesApi",
    "apiConsumedBy",
  ],
  [
    "apiProvidedBy",
    "providesApi",
  ],
  [
    "hasPart",
    "partOf",
  ],
  [
    "parentOf",
    "childOf",
  ],
  [
    "hasMember",
    "memberOf",
  ],
  [
    "dependsOn",
    "dependencyOf",
  ],
  [
    "testOf",
    "testBy",
  ],
]
`;

exports[`useRelationsCatalogGraphDefaults should return relationPairs overwrite 1`] = `
[
  [
    "overwriteOf",
    "overwriteBy",
  ],
]
`;
